// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package sai;

// A switch can provide limitations of the WCMP groups that can be installed
// on the switch. A switch may report a ResourceExhausted error, if:
//  1) The sum of all weights in a group exceeds `max_total_weight_per_group`.
//  2) The number of distinct weights in a group exceeds
//     `max_distinct_weights_per_group`.
message WcmpGroupLimitations {
  // The maximum total weight (sum of weights) within a single group.
  // For example, a max of 10:
  //  - PERMITs a group with weights {1, 2, 3, 4}
  //  - REJECTs a group with weights {11}
  // This limit is applied in addition to other limits that use different
  // semantics (such as max members per group).
  //
  // The default value of 0 means there is no limit.
  int32 max_total_weight_per_group = 1;

  // The maximum number of distinct weights allowed within a single group.
  // As an example, a max of 3:
  //  - PERMITs a group with weights {1, 8, 32, 32}
  //  - REJECTs a group with weights {1, 2, 3, 4}
  //
  // The default value of 0 means there is no limit.
  int32 max_distinct_weights_per_group = 2;
}

// TODO: Extend p4.v1.CapabilitiesResponse to include
// ResourceCapabilities.
// Resource constraints and availability reported by the switch.
// By experimental, we mean that the P4Runtime specification does not provide a
// standardized way to report these capabilities yet.
message ExperimentalResourceCapabilities {
  // Limitations of WCMP groups.
  WcmpGroupLimitations wcmp_group_limitations = 1;
}
